### PR TITLE
[REVIEW] Update shapefile_readers.hpp to libcudf++

### PR DESCRIPTION
**Pull request by cwharris**
Contributes to #154.

### CODE REVIEW

**Comment by GPUtester** [comment]
Please update the changelog in order to start CI tests.

View the gpuCI docs [here](https://docs.rapids.ai/gpuci#open-pr-or-update-pr).

File: cpp/include/cuspatial/shapefile_reader.hpp [hunk 1]
```diff
@@ -0,0 +1,47 @@
...
+#include <cudf/column/column.hpp>
+#include <cudf/types.hpp>
+
+namespace cuspatial {
+
+/**
+ * @brief read polygon data from an ESRI Shapefile.
+ *
+ * @param[in] filename: ESRI Shapefile file path (usually ends in .shp)
+ * @param[in] mr:       Optional, The resource to use for all allocations
```
Comment (harrism) [inline cpp/include/cuspatial/shapefile_reader.hpp:28]:
```suggestion
 * @param[in] mr:       Optional, The resource to use to allocate the returned data
```

Comment (harrism) [reply cpp/include/cuspatial/shapefile_reader.hpp:28]: Only output data should be allocated with mr. Internal temporary allocations should use the default resource.

File: cpp/src/io/shp/polygon_shapefile_reader.cu [hunk 2]
```diff
@@ -0,0 +1,103 @@
...
+
+#include <memory>
+#include <string>
+
+namespace cuspatial {
+namespace detail {
+
+template<typename T>
+std::unique_ptr<cudf::column>
+make_column(std::vector<T> source, cudaStream_t stream, rmm::mr::device_memory_resource* mr)
```
Comment (harrism) [inline cpp/src/io/shp/polygon_shapefile_reader.cu:37]: Hmmm, seems like we are missing a make_column_from_vector factory in cuDF?

File: cpp/src/io/shp/polygon_shapefile_reader.cpp [hunk 3]
```diff
@@ -0,0 +1,161 @@
...
+
+namespace {
+
+cudf::size_type read_ring(OGRLinearRing const& ring,
+                          std::vector<double>& xs,
+                          std::vector<double>& ys)
+{
+    cudf::size_type num_vertices = ring.getNumPoints();
+
+    // append points in reverse order
```
Comment (harrism) [inline cpp/src/io/shp/polygon_shapefile_reader.cpp:34]: Why in reverse order?

Comment (cwharris) [reply cpp/src/io/shp/polygon_shapefile_reader.cpp:34]: Good question. That was the original implementation, I believe. I’ll see if I can find out _why_, though.

Comment (cwharris) [reply cpp/src/io/shp/polygon_shapefile_reader.cpp:34]: @zhangjianting may know the answer.

File: cpp/src/io/shp/polygon_shapefile_reader.cu (hunk 2 above)
Comment (cwharris) [reply cpp/src/io/shp/polygon_shapefile_reader.cu:37]: Opened rapidsai/cudf#5150 to track this.

File: cpp/include/cuspatial/shapefile_reader.hpp [hunk 4]
```diff
@@ -0,0 +1,47 @@
...
+ *  2 - FLOAT64: x component of polygon points
+ *  3 - FLOAT64: y component of polygon points
+ *
+ * @note The number of polygons is equal to the length of the first column
+ *
+**/
+std::tuple<std::unique_ptr<cudf::column>,
+           std::unique_ptr<cudf::column>,
+           std::unique_ptr<cudf::column>,
+           std::unique_ptr<cudf::column>>
```
Comment (trxcllnt) [inline cpp/include/cuspatial/shapefile_reader.hpp:42]: I just remembered we can't read tuples of non-copyable types in Cython, so we need to update this to return a `vector<unique_ptr<column>>`

Comment (trxcllnt) [reply cpp/include/cuspatial/shapefile_reader.hpp]: Done.

**Comment by trxcllnt** [comment]
Remembered Cython can't unpack `std::tuple` of non-copyable types like `unique_ptr`. I changed the `read_polygon_shapefile` C++ return type to `std::vector<std::unique_ptr<cudf::column>>`, all good now.

**Comment by cwharris** [comment]
@trxcllnt That applies to tuples, in general, but actually cython doesn't have to unpack _pairs_, we can move them individually as `.first` and`.last` instead.

**Pull request merged**